### DISCUSSION TITLE
How to use arabert-v2 completely offline?

### DISCUSSION THREAD

**Discussion by atytxt** [aub-mind/arabert#127] [Q&A]
Hi

i want to use the arabert-v2 preprocess offline. but it seems saving model is not enough for arabert preprocess function. What else should I download? any ideas?

**Comment by WissamAntoun** [comment]
The preprocessor for Arabert v2 and v1, uses the Farasa segmenter package.
Now the farasa segmenter python package internally calls a java .jar file to do the segmentation.

In order to use the farasa segmenter package offline, I think you need to clone the project from here https://github.com/MagedSaeed/farasapy . Then modify the target location where the zip file containing the farasa jars to a location where you have manually already downloaded the zip file.

This is a complicated task, which can be made easier if you have a one-time connection to let the farasapy package download and setup the files on its own.

I suggest opening a discussion or issue in farasapy and further discuss it there.

**Reply by atytxt** [reply]
Thank you, but another problem. when I save bert-base-arabertv2 model, then as model name i give the path but I get this error:
WARNING:root:Model provided is not in the accepted model list. Preprocessor will default to a base Arabic preprocessor

my code for saving model:
from transformers import AutoTokenizer, AutoModel

tokenizer = AutoTokenizer.from_pretrained("aubmindlab/bert-base-arabertv2")
model = AutoModel.from_pretrained("aubmindlab/bert-base-arabertv2")
tokenizer.save_pretrained('/content/gdrive/MyDrive/st_model/')
model.save_pretrained('/content/gdrive/MyDrive/st_model/')

**Reply by WissamAntoun** [reply]
the code for saving the model isn't saving the preprocessor object. 

Also when you re-create the preprocessor are you giving it "aubmindlab/bert-base-arabertv2" or the path to the saved model?

**Reply by atytxt** [reply]
I give the path to the saved model.

'the code for saving the model isn't saving the preprocessor object.'
so what should I do about it?

**Reply by WissamAntoun** [reply]
you don't have to save the preprocessor. Just give it the string "aubmindlab/bert-base-arabertv2" since it was the one you used for training.

**Reply by atytxt** [reply]
I want to use it for preprocessing inputs of some pretrained classification models. I mean I do not train anything in my code.

**Reply by WissamAntoun** [reply]
Then either always give it one of the accepted model list.
Or you can set the flags just like you want without even passing a model_name

**Reply by atytxt** [reply]
It means it always needs internet connection? or once it download it the first time, it is possible to use it offline?

**Reply by WissamAntoun** [reply]
Only need to connect and download once

**Comment by atytxt** [comment]
so another issue is that normalizing a sentence with basic models takes few seconds(or maybe less than a second) but normalizing a single sentence with 'aubmindlab/bert-base-arabertv2' takes more than 1 minute. Any idea or suggestion?

**Reply by WissamAntoun** [reply]
Weird it should take few milliseconds even when doing segmentation. What hardware are you using?

**Reply by atytxt** [reply]
even with keep_emojis= True?

**Reply by WissamAntoun** [reply]
kepp_emojis makes it slower but not by that much i say less than 100ms per normal sentence.
